src/main/java/org/candlepin/resource/ConsumerResource.java

```diff
@@ -1063,15 +1064,30 @@ public Consumer createConsumerFromDTO(ConsumerDTO consumer, ConsumerType type, P
         }
     }
 
-    private List<ActivationKey> findActivationKeys(
+    /**
+     * We need to make sure the requested keys are available. In addition, the order in
+     *  which they are specified in the call should be maintained for their later application.
+     *
+     * @param principal
+     * @param owner
+     * @param keyStrings
+     * @return
+     */
+    private List<ActivationKey> findActivationKeysInOrder(
         Principal principal, Owner owner, Set<String> keyStrings) {
         if (keyStrings.isEmpty()) {
             return Collections.emptyList();
         }
 
         List<ActivationKey> foundKeys = this.activationKeyCurator.findByKeyNames(owner.getKey(), keyStrings);
-
-        Set<String> found = foundKeys.stream()
+        List<String> sortOrder = new ArrayList<>();
```

Comment:
This block of collection juggling is unnecessary. The entirety of it can be replaced by a single call to `List.sort`:

```
List<String> keyNames = List.of(keyString.split("\s*,\*"));
List<ActivationKey> keys = this.activationKeyCurator.findByKeyNames(...);

// insert validation of resolved keys here

keys.sort(Comparator.comparing(key -> keyNames.indexOf(key.getName()));

return keys;
```

Note that we also need to properly handle the case where a given activation key is specified multiple times, and our principal check and general warning output is incorrect as well. We should always be warning when an AK is specified that doesn't exist (it's not user facing anyway), and if the intent is that non-AKPrincipals can't use AKs, then we shouldn't do *any* processing of fetched AKs in those cases.

Reply:
The intent appears to be that we only log if it is a non-AK principal. I just dropped the condition.